README.md
# focus-group

[![Build Status](https://travis-ci.org/davidtheclark/focus-group.svg?branch=master)](https://travis-ci.org/davidtheclark/focus-group)
[![Coverage Status](https://coveralls.io/repos/github/davidtheclark/focus-group/badge.svg?branch=master)](https://coveralls.io/github/davidtheclark/focus-group?branch=master)

Create a group of nodes with special focus-related powers.

Specifically, you can do the following with your focus group:

- Use different keybindings to move focus through the nodes
- Type (with letters) to jump focus to a specific node based on its text

Essentially, it mimics some of the essential keyboard interactions of a native `<select>`.

These kinds of powers are useful for:

- Accessible menus, like [react-aria-menubutton](//github.com/davidtheclark/react-aria-menubutton)
- Any other widgets whose keyboard UX will improve by enabling arrow-key navigation and letter-key jumping

## Concepts

A focus-group is composed of members.

The order of the members matters, because focus moves forwards and backwards through the group, in order.

Each member consists of a DOM node and some text associated with that node.
The member's text will be used for letter-key jumping (a.k.a. string searching).
Each member's text can be established in a few ways:

- It can be manually specified when adding the member to the group, via `setMembers()` or `addMember()` (see below).
- If the member's node has a `data-focus-group-text` attribute, that value will serve as the member's text.
- If neither of the above is provided, the member's text will be the `textContent` of its node.

## Keyboard Interactions

When focus is inside the focus-group, the following things should happen:

- If you press one of your `next` `keybindings` (the down arrow by default), focus moves
  from the currently focused member to the next member in the group (or wraps back
  to the front, according to the `wrap` option).
- If you press one of your `prev` `keybindings` (the up arrow by default), focus moves
  from the currently focused member to the previous member in the group (or wraps around
  to the back, according to the `wrap` option).
- If you press a letter key, string searching begins (see below!).

### String searching

If the option `stringSearch` is `true` and focus is within the group, the following things happen:

- When you start typing, focus moves to the first member whose registered text begins with
  whatever you've been typing.
- As long as each keystroke occurs within `stringSearchDelay`,
  the search string will extend (e.g. `f` -> `fa` -> `far` -> `farm`) and focus will move
  accordingly.
- If no text matches the search string, focus will not move.
- After you have not typed any letters for `stringSearchDelay`, the search
  string resets and you can start over (e.g. you type `fa` then wait and type `go` to match `gorge`).

This all mimics the native `<select>` behavior.

Note that like the native `<select>`, typing only matches the *beginning of words*. So you can't focus `David Clark` by typing `Clark`.

## API

### var focusGroup = createFocusGroup([options])

This is the function you get when you `require()` or `import` the module.

```js
var createFocusGroup = require('focus-group');
var myMegaMenuFocusGroup = createFocusGroup();
```

#### Options

**members** { Array }: Designate initial members of the group. Can be any of the following:

- An array of DOM nodes (or a NodeList, like what's returned by `querySelectorAll()`)
- An array of member objects, each object with the following properties: `node` (the DOM node),
  and (optionally) `text` (the text that should be associated with that node for letter-navigation)

You can omit this option and add members later with `addMember()` or `setMembers()`. Default: `[]`.

**keybindings** { Object of `'next'`, `'prev'`, `'first'`, or `'last'` }:
Specify which key events should move the focus *forward*, *back*, to the *first* member, or to the *last* member through the group. Provide objects (or arrays of objects) that describe the requirements of a [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent) that should trigger that keybinding.

Default:
``` js
{
  next: { keyCode: 40 }, // ArrowDown
  prev: { keyCode: 38 }, // ArrowUp
}
```

**wrap** { Boolean }:
If `true`, when the arrow keys are moving focus they will wrap around the group. That is, when focus is on the last item and you move focus forward, the first item will focus; and when focus is on the first item and you move focus back, the last item will focus.

**stringSearch** { Boolean }:
If `true`, string searching is enabled (see below).
Default: `false`.

**stringSearchDelay** { Number }:
The number of milliseconds that should elapse between the user's last letter entry (with the keyboard)
and a refresh of the string search (see below).
Default: `800`.

### focusGroup.activate()

Start this group listening to keyboard events and responding accordingly.

Returns the focus group instance.

### focusGroup.deactivate()

Stop this group listening to keyboard events.

Returns the focus group instance.

### focusGroup.addMember(member[, index])

Add a member to the group.

`member` can be any of the following:

- A DOM node
- An object with the following properties:
  - `node` (the node itself)
  - (optionally) `text`: Text that should be associated with that node for letter-navigation. If none is provided, focus-group will check for a `data-focus-group-text` attribute or fallback to the node's `textContent.`

If `index` is provided, the member will be added at that index.
Otherwise, it will be added to the end of the group.

Returns the focus group instance.

### focusGroup.removeMember(member)

Remove a member from the group.

`member` can be any of the following:

- A DOM node
- An index for the member that should be removed.

Returns the focus group instance.

### focusGroup.clearMembers()

Empty the focus group of members.

Returns the focus group instance.

### focusGroup.setMembers(members)

Set the focus group's members (clearing any that already exist).

`members` can be any of the following:

- An array of DOM nodes (or a NodeList, like what's returned by `querySelectorAll()`)
- An array of member objects, each object with the following properties:
  - `node` (the node itself)
  - (optionally) `text`: Text that should be associated with that node for letter-navigation. If none is provided, focus-group will check for a `data-focus-group-text` attribute or fallback to the node's `textContent.

Returns the focus group instance.

### focusGroup.getMembers()

Returns the focus group's current array of members.

Each item in the array is an object with `node` and `text` properties.

### focusGroup.focusNodeAtIndex(index)

Focuses the node at a particular index in the focus group's member array.

If no member exists at that index, does nothing.

Returns the focus group instance.

### focusGroup.moveFocusForward()

Moves the focus forward one member, if focus is already within the group.

If focus is not within the group, does nothing.

Returns the index of the newly focused member.

### focusGroup.moveFocusBack()

Moves the focus back one member, if focus is already within the group.

If focus is not within the group, does nothing.

Returns the index of the newly focused member.

## Contributing

Please note that this project is released with a Contributor Code of Conduct.
By participating in this project you agree to abide by its terms.


CHANGELOG.md
# Changelog

## 0.3.0
- Removed `forwardArrows` and `backArrows` options in favor of `keybindings` option that accepts `next`, `prev`, `first`, or `last`.

## 0.2.3
- Fix bug where providing a `0` index to `addMember()` failed.

## 0.2.2
- Fix bug where `setMembers()` wasn't calling `clearMembers()` first.

## 0.2.1
- Fix bug when there's `member.node` but no `member.text`.

## 0.2.0

- Initial release.


test/index.js
import test from 'ava';
import sinon from 'sinon';
import createFocusGroup from '..';

test.beforeEach(t => {
  t.context.mockDocument = {
    addEventListener: sinon.spy(),
    removeEventListener: sinon.spy(),
  };
  t.context.defaultTestOptions = {
    _document: t.context.mockDocument,
  };
});

test('FocusGroup#constructor with default settings', t => {
  const noOptionsGroup = createFocusGroup(t.context.defaultTestOptions);
  t.same(noOptionsGroup._settings, {
    forwardArrows: ['down'],
    backArrows: ['up'],
    letterNavigation: undefined,
    cycle: undefined,
  });
  t.same(noOptionsGroup._nodes, []);
});

test('FocusGroup#_getActiveNodeIndex', t => {
  // Use numbers instead of nodes for simplicity
  t.context.mockDocument.activeElement = 3;
  var focusGroup = createFocusGroup(t.context.defaultTestOptions);
  focusGroup._nodes = [1, 2, 3, 4, 5];
  t.is(focusGroup._getActiveNodeIndex(), 2);
});

test('FocusGroup#activate, without already active group', t => {
  const focusGroup = createFocusGroup(t.context.defaultTestOptions);
  const activateReturnValue = focusGroup.activate();
  t.is(t.context.mockDocument.addEventListener.callCount, 1);
  t.same(t.context.mockDocument.addEventListener.getCall(0).args, [
    'keydown',
    focusGroup._handleKeyDown,
    true,
  ]);
  t.is(activateReturnValue, focusGroup);
  t.is(focusGroup.constructor.activeGroup, focusGroup);
});

test('FocusGroup#active on a second group deactivates the first', t => {
  const focusGroupA = createFocusGroup(t.context.defaultTestOptions);
  const focusGroupADeactivate = sinon.spy(focusGroupA, 'deactivate');
  const focusGroupB = createFocusGroup(t.context.defaultTestOptions);
  focusGroupA.activate();
  focusGroupB.activate();
  t.is(focusGroupADeactivate.callCount, 1);
  t.is(focusGroupA.constructor.activeGroup, focusGroupB);
});

test('FocusGroup#deactivate', t => {
  const focusGroup = createFocusGroup(t.context.defaultTestOptions);
  focusGroup.activate();
  const deactivateReturnValue = focusGroup.deactivate();
  t.is(t.context.mockDocument.removeEventListener.callCount, 1);
  t.same(t.context.mockDocument.removeEventListener.getCall(0).args, [
    'keydown',
    focusGroup._handleKeyDown,
    true,
  ]);
  t.is(deactivateReturnValue, focusGroup);
  t.is(focusGroup.constructor.activeGroup, null);
});

test('FocusGroup#handleKeyDown with down arrow, default settings, no activeElement', t => {
  const focusGroup = createFocusGroup(t.context.defaultTestOptions);
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockDownArrowEvent = createMockEvent('ArrowDown');
  focusGroup.handleKeyDown(mockDownArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(mockDownArrowEvent.preventDefault.callCount, 0);
});

test('FocusGroup#handleKeyDown with down arrow, default settings, and an activeElement outside the group', t => {
  t.context.mockDocument.activeElement = 7;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockDownArrowEvent = createMockEvent('ArrowDown');
  focusGroup.handleKeyDown(mockDownArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(mockDownArrowEvent.preventDefault.callCount, 0);
});

test('FocusGroup#handleKeyDown with down arrow, initialNodes, and activeElement inside the group', t => {
  t.context.mockDocument.activeElement = 4;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockDownArrowEvent = createMockEvent('ArrowDown');
  focusGroup.handleKeyDown(mockDownArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 1);
  t.same(focusGroup.moveFocusForward.getCall(0).args, [1],
    'moveFocusForward called with active node index');
  t.is(mockDownArrowEvent.preventDefault.callCount, 1);
});

test('FocusGroup#handleKeyDown with keyCode 40, initialNodes, and activeElement inside the group', t => {
  t.context.mockDocument.activeElement = 4;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockDownArrowEvent = createMockEvent(40);
  focusGroup.handleKeyDown(mockDownArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 1);
  t.same(focusGroup.moveFocusForward.getCall(0).args, [1],
    'moveFocusForward called with active node index');
  t.is(mockDownArrowEvent.preventDefault.callCount, 1);
});

test('FocusGroup#handleKeyDown with up arrow, default settings, no activeElement', t => {
  const focusGroup = createFocusGroup(t.context.defaultTestOptions);
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockUpArrowEvent = createMockEvent('ArrowUp');
  focusGroup.handleKeyDown(mockUpArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(mockUpArrowEvent.preventDefault.callCount, 0);
});

test('FocusGroup#handleKeyDown with up arrow, default settings, and an activeElement outside the group', t => {
  t.context.mockDocument.activeElement = 7;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockUpArrowEvent = createMockEvent('ArrowUp');
  focusGroup.handleKeyDown(mockUpArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(mockUpArrowEvent.preventDefault.callCount, 0);
});

test('FocusGroup#handleKeyDown with up arrow, initialNodes, and activeElement inside the group', t => {
  t.context.mockDocument.activeElement = 4;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockUpArrowEvent = createMockEvent('ArrowUp');
  focusGroup.handleKeyDown(mockUpArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 1);
  t.same(focusGroup.moveFocusBack.getCall(0).args, [1],
    'moveFocusBack called with active node index');
  t.is(mockUpArrowEvent.preventDefault.callCount, 1);
});

test('FocusGroup#handleKeyDown with keyCode 38, initialNodes, and activeElement inside the group', t => {
  t.context.mockDocument.activeElement = 4;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockUpArrowEvent = createMockEvent(38);
  focusGroup.handleKeyDown(mockUpArrowEvent);
  t.is(focusGroup.moveFocusByLetter.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(focusGroup.moveFocusBack.callCount, 1);
  t.same(focusGroup.moveFocusBack.getCall(0).args, [1],
    'moveFocusBack called with active node index');
  t.is(mockUpArrowEvent.preventDefault.callCount, 1);
});

test('FocusGroup#handleKeyDown with keyCode 65 (letter A), default settings, and an activeElement inside the group', t => {
  t.context.mockDocument.activeElement = 6;
  const focusGroup = createFocusGroup(Object.assign(t.context.defaultTestOptions, {
    initialNodes: [3, 4, 5, 6],
  }));
  focusGroup.moveFocusByLetter = sinon.spy();
  focusGroup.moveFocusBack = sinon.spy();
  focusGroup.moveFocusForward = sinon.spy();
  const mockLetterKeyEvent = createMockEvent(65);
  focusGroup.handleKeyDown(mockLetterKeyEvent);
  t.is(focusGroup.moveFocusBack.callCount, 0);
  t.is(focusGroup.moveFocusForward.callCount, 0);
  t.is(focusGroup.moveFocusByLetter.callCount, 1);
  t.same(focusGroup.moveFocusByLetter.getCall(0).args, [mockLetterKeyEvent],
    'moveFocusByLetter called with the event'););
});

// test('FocusGroup#moveFocusForward' t => {
//   const focusGroup = createFocusGroup(t.context.defaultTestOptions);
//   const mockNode = {};
//   const focusNodeAtIndexStub = stub(focusGroup, 'focusNodeAtIndex', i => {
//     if (i === 6) return mockNode;
//   });
//   focusGroup.moveFocusForward()
// });

function createMockEvent(keyInfo) {
  const event = {
    preventDefault: sinon.spy(),
  };
  if (typeof keyInfo === 'number') {
    event.keyCode = keyInfo;
  } else {
    event.key = keyInfo;
  }
  return event;
}


index.js
function FocusGroup(options) {
  options = options || {};
  options.keybindings = options.keybindings || {};
  this._settings = {
    keybindings: {
      next: (options.keybindings.next) || { keyCode: 40 },
      prev: (options.keybindings.prev) || { keyCode: 38 },
      first: options.keybindings.first,
      last: options.keybindings.last,
    },
    wrap: options.wrap,
    stringSearch: options.stringSearch,
    stringSearchDelay: 800
  };

  // Construct a keybinding lookup that will be more useful later
  this._keybindingsLookup = [];
  var action;
  var eventMatchers
  for (action in this._settings.keybindings) {
    eventMatchers = this._settings.keybindings[action];
    if (!eventMatchers) continue;
    [].concat(eventMatchers).forEach(function(eventMatcher) {
      eventMatcher.metaKey = eventMatcher.metaKey || false;
      eventMatcher.ctrlKey = eventMatcher.ctrlKey || false;
      eventMatcher.altKey = eventMatcher.altKey || false;
      eventMatcher.shiftKey = eventMatcher.shiftKey || false;
      this._keybindingsLookup.push({
        action: action,
        eventMatcher: eventMatcher
      });
    }.bind(this));
  }

  this._searchString = '';
  this._members = [];
  if (options.members) this.setMembers(options.members);
  this._boundHandleKeydownEvent = this._handleKeydownEvent.bind(this);
}

FocusGroup.prototype.activate = function() {
  // Use capture in case other libraries might grab it first -- i.e. React
  document.addEventListener('keydown', this._boundHandleKeydownEvent, true);
  return this;
};

FocusGroup.prototype.deactivate = function() {
  document.removeEventListener('keydown', this._boundHandleKeydownEvent, true);
  this._clearSearchStringRefreshTimer();
  return this;
};

FocusGroup.prototype._handleKeydownEvent = function(event) {
  // Only respond to keyboard events when
  // focus is already within the focus-group
  var activeElementIndex = this._getActiveElementIndex();
  if (activeElementIndex === -1) return;

  // See if the event matches any registered keybinds
  var eventBound = false;
  this._keybindingsLookup.forEach(function(keybinding) {
    if (!matchesEvent(keybinding.eventMatcher, event)) return;
    eventBound = true;
    event.preventDefault();
    switch (keybinding.action) {
      case 'next':
        this.moveFocusForward();
        break;
      case 'prev':
        this.moveFocusBack();
        break;
      case 'first':
        this.moveFocusToFirst();
        break;
      case 'last':
        this.moveFocusToLast();
        break;
      default: return;
    }
  }.bind(this));

  if (!eventBound) {
    this._handleUnboundKey(event);
  }
};

FocusGroup.prototype.moveFocusForward = function() {
  var activeElementIndex = this._getActiveElementIndex();
  var targetIndex;
  if (activeElementIndex < this._members.length - 1) {
    targetIndex = activeElementIndex + 1;
  } else if (this._settings.wrap) {
    targetIndex = 0;
  } else {
    targetIndex = activeElementIndex;
  }
  this.focusNodeAtIndex(targetIndex);
  return targetIndex;
};

FocusGroup.prototype.moveFocusBack = function() {
  var activeElementIndex = this._getActiveElementIndex();
  var targetIndex;
  if (activeElementIndex > 0) {
    targetIndex = activeElementIndex - 1;
  } else if (this._settings.wrap) {
    targetIndex = this._members.length - 1;
  } else {
    targetIndex = activeElementIndex;
  }
  this.focusNodeAtIndex(targetIndex);
  return targetIndex;
};

FocusGroup.prototype.moveFocusToFirst = function() {
  this.focusNodeAtIndex(0);
};

FocusGroup.prototype.moveFocusToLast = function() {
  this.focusNodeAtIndex(this._members.length - 1);
};

FocusGroup.prototype._handleUnboundKey = function(event) {
  if (!this._settings.stringSearch) return;

  // While a string search is underway, ignore spaces
  // and prevent the default space-key behavior
  if (this._searchString !== '' && (event.key === ' ' || event.keyCode === 32)) {
    event.preventDefault();
    return -1;
  }

  // Only respond to letter keys
  if (!isLetterKeyCode(event.keyCode)) return -1;

  // If the letter key is part of a key combo,
  // let it do whatever it was going to do
  if (event.ctrlKey || event.metaKey || event.altKey) return -1;

  event.preventDefault();

  this._addToSearchString(String.fromCharCode(event.keyCode));
  this._runStringSearch();
};

FocusGroup.prototype._clearSearchString = function() {
  this._searchString = '';
};

FocusGroup.prototype._addToSearchString = function(letter) {
  // Always store the lowercase version of the letter
  this._searchString += letter.toLowerCase();
};

FocusGroup.prototype._startSearchStringRefreshTimer = function() {
  var self = this;
  this._clearSearchStringRefreshTimer();
  this._stringSearchTimer = setTimeout(function() {
    self._clearSearchString();
  }, this._settings.stringSearchDelay);
};

FocusGroup.prototype._clearSearchStringRefreshTimer = function() {
  clearTimeout(this._stringSearchTimer);
};

FocusGroup.prototype._runStringSearch = function() {
  this._startSearchStringRefreshTimer();
  this.moveFocusByString(this._searchString);
};

FocusGroup.prototype.moveFocusByString = function(str) {
  var member;
  for (var i = 0, l = this._members.length; i < l; i++) {
    member = this._members[i];
    if (!member.text) continue;

    if (member.text.indexOf(str) === 0) {
      return focusNode(member.node);
    }
  }
};

FocusGroup.prototype._findIndexOfNode = function(searchNode) {
  for (var i = 0, l = this._members.length; i < l; i++) {
    if (this._members[i].node === searchNode) {
      return i;
    }
  }
  return -1;
};

FocusGroup.prototype._getActiveElementIndex = function() {
  return this._findIndexOfNode(document.activeElement);
};

FocusGroup.prototype.focusNodeAtIndex = function(index) {
  var member = this._members[index];
  if (member) focusNode(member.node);
  return this;
};

FocusGroup.prototype.addMember = function(memberData, index) {
  var node = memberData.node || memberData;
  var nodeText = memberData.text || node.getAttribute('data-focus-group-text') || node.textContent || '';

  this._checkNode(node);

  var cleanedNodeText = nodeText.replace(/[\W_]/g, '').toLowerCase();
  var member = {
    node: node,
    text: cleanedNodeText,
  };

  if (index !== null && index !== undefined) {
    this._members.splice(index, 0, member);
  } else {
    this._members.push(member);
  }
  return this;
};

FocusGroup.prototype.removeMember = function(member) {
  var removalIndex = (typeof member === 'number')
    ? member
    : this._findIndexOfNode(member);
  if (removalIndex === -1) return;
  this._members.splice(removalIndex, 1);
  return this;
};

FocusGroup.prototype.clearMembers = function() {
  this._members = [];
  return this;
};

FocusGroup.prototype.setMembers = function(nextMembers) {
  this.clearMembers();
  for (var i = 0, l = nextMembers.length; i < l; i++) {
    this.addMember(nextMembers[i]);
  }
  return this;
};

FocusGroup.prototype.getMembers = function() {
  return this._members;
};

FocusGroup.prototype._checkNode = function(node) {
  if (!node.nodeType || node.nodeType !== window.Node.ELEMENT_NODE) {
    throw new Error('focus-group: only DOM nodes allowed');
  }
  return node;
};

function matchesEvent(matcher, event) {
  for (var key in matcher) {
    if (event[key] !== undefined && matcher[key] !== event[key]) return false;
  }
  return true;
}

function isLetterKeyCode(keyCode) {
  return keyCode >= 65 && keyCode <= 90;
}

function focusNode(node) {
  if (!node || !node.focus) return;
  node.focus();
  if (node.tagName.toLowerCase() === 'input') node.select();
}

module.exports = function createFocusGroup(options) {
  return new FocusGroup(options);
};


demo/demo-bundle.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
var createFocusGroup = require('..');

var stateNodes = [].slice.call(document.querySelectorAll('.state'));
createFocusGroup({
	members: stateNodes,
	stringSearch: true,
}).activate();

var wraNodes = document.querySelectorAll('.wra');
createFocusGroup({
	members: wraNodes,
	stringSearch: true,
	forwardArrows: ['right', 'down'],
	backArrows: ['left', 'up'],
	wrap: true,
}).activate();

},{"..":2}],2:[function(require,module,exports){
function FocusGroup(options) {
  options = options || {};
  this._settings = {
    forwardArrows: options.forwardArrows || ['down'],
    backArrows: options.backArrows || ['up'],
    wrap: options.wrap,
    stringSearch: options.stringSearch,
    stringSearchDelay: 800,
  };
  this._searchString = '';
  this._members = [];
  if (options.members) this.setMembers(options.members);
  this._boundHandleKeydownEvent = this._handleKeydownEvent.bind(this);
}

FocusGroup.prototype.activate = function() {
  // Use capture in case other libraries might grab it first -- i.e. React
  document.addEventListener('keydown', this._boundHandleKeydownEvent, true);
  return this;
};

FocusGroup.prototype.deactivate = function() {
  document.removeEventListener('keydown', this._boundHandleKeydownEvent, true);
  this._clearSearchStringRefreshTimer();
  return this;
};

FocusGroup.prototype._handleKeydownEvent = function(event) {
  // Only respond to keyboard events when
  // focus is already within the focus-group
  var activeElementIndex = this._getActiveElementIndex();
  if (activeElementIndex === -1) return;

  var arrow = getEventArrowKey(event);

  if (!arrow) {
    this._handleNonArrowKey(event);
    return;
  }

  if (this._settings.forwardArrows.indexOf(arrow) !== -1) {
    event.preventDefault();
    this.moveFocusForward();
    return;
  }

  if (this._settings.backArrows.indexOf(arrow) !== -1) {
    event.preventDefault();
    this.moveFocusBack();
    return;
  }
};

FocusGroup.prototype.moveFocusForward = function() {
  var activeElementIndex = this._getActiveElementIndex();
  var targetIndex;
  if (activeElementIndex < this._members.length - 1) {
    targetIndex = activeElementIndex + 1;
  } else if (this._settings.wrap) {
    targetIndex = 0;
  } else {
    targetIndex = activeElementIndex;
  }
  this.focusNodeAtIndex(targetIndex);
  return targetIndex;
};

FocusGroup.prototype.moveFocusBack = function() {
  var activeElementIndex = this._getActiveElementIndex();
  var targetIndex;
  if (activeElementIndex > 0) {
    targetIndex = activeElementIndex - 1;
  } else if (this._settings.wrap) {
    targetIndex = this._members.length - 1;
  } else {
    targetIndex = activeElementIndex;
  }
  this.focusNodeAtIndex(targetIndex);
  return targetIndex;
};

FocusGroup.prototype._handleNonArrowKey = function(event) {
  if (!this._settings.stringSearch) return;

  // While a string search is underway, ignore spaces
  // and prevent the default space-key behavior
  if (this._searchString !== '' && (event.key === ' ' || event.keyCode === 32)) {
    event.preventDefault();
    return -1;
  }

  // Only respond to letter keys
  if (!isLetterKeyCode(event.keyCode)) return -1;

  // If the letter key is part of a key combo,
  // let it do whatever it was going to do
  if (event.ctrlKey || event.metaKey || event.altKey) return -1;

  event.preventDefault();

  this._addToSearchString(String.fromCharCode(event.keyCode));
  this._runStringSearch();
};

FocusGroup.prototype._clearSearchString = function() {
  this._searchString = '';
};

FocusGroup.prototype._addToSearchString = function(letter) {
  // Always store the lowercase version of the letter
  this._searchString += letter.toLowerCase();
};

FocusGroup.prototype._startSearchStringRefreshTimer = function() {
  var self = this;
  this._clearSearchStringRefreshTimer();
  this._stringSearchTimer = setTimeout(function() {
    self._clearSearchString();
  }, this._settings.stringSearchDelay);
};

FocusGroup.prototype._clearSearchStringRefreshTimer = function() {
  clearTimeout(this._stringSearchTimer);
}

FocusGroup.prototype._runStringSearch = function() {
  this._startSearchStringRefreshTimer();
  this.moveFocusByString(this._searchString);
}

FocusGroup.prototype.moveFocusByString = function(str) {
  var member;
  for (var i = 0, l = this._members.length; i < l; i++) {
    member = this._members[i];
    if (!member.text) continue;

    if (member.text.indexOf(str) === 0) {
      return focusNode(member.node);
    }
  }
}

FocusGroup.prototype._findIndexOfNode = function(searchNode) {
  for (var i = 0, l = this._members.length; i < l; i++) {
    if (this._members[i].node === searchNode) {
      return i;
    }
  }
  return -1;
}

FocusGroup.prototype._getActiveElementIndex = function() {
  return this._findIndexOfNode(document.activeElement);
};

FocusGroup.prototype.focusNodeAtIndex = function(index) {
  var member = this._members[index];
  if (member) focusNode(member.node);
  return this;
};

FocusGroup.prototype.addMember = function(member, index) {
  var node;
  var nodeText;
  if (member.node && member.text) {
    node = this._checkNode(member.node);
    nodeText = member.text;
  } else {
    node = this._checkNode(member);
    nodeText = node.getAttribute('data-focus-group-text') || node.textContent || '';
  }

  var cleanedNodeText = nodeText.replace(/[\W_]/g, '').toLowerCase();
  var member = {
    node: node,
    text: cleanedNodeText,
  };

  if (index) {
    this._members.splice(index, 0, member);
  } else {
    this._members.push(member);
  }
  return this;
};

FocusGroup.prototype.removeMember = function(member) {
  var removalIndex = (typeof member === 'number')
    ? member
    : this._findIndexOfNode(member);
  if (removalIndex === -1) return;
  this._members.splice(removalIndex, 1);
  return this;
};

FocusGroup.prototype.clearMembers = function() {
  this._members = [];
  return this;
};

FocusGroup.prototype.setMembers = function(nextMembers) {
  for (var i = 0, l = nextMembers.length; i < l; i++) {
    this.addMember(nextMembers[i]);
  }
  return this;
};

FocusGroup.prototype.getMembers = function() {
  return this._members;
};

FocusGroup.prototype._checkNode = function(node) {
  if (!node.nodeType || node.nodeType !== window.Node.ELEMENT_NODE) {
    throw new Error('focus-group: only DOM nodes allowed');
  }
  return node;
};

function getEventArrowKey(event) {
  if (event.key === 'ArrowUp' || event.keyCode === 38) return 'up';
  if (event.key === 'ArrowDown' || event.keyCode === 40) return 'down';
  if (event.key === 'ArrowLeft' || event.keyCode === 37) return 'left';
  if (event.key === 'ArrowRight' || event.keyCode === 39) return 'right';
  return null;
}

function isLetterKeyCode(keyCode) {
  return keyCode >= 65 && keyCode <= 90;
}

function focusNode(node) {
  if (!node || !node.focus) return;
  node.focus();
  if (node.tagName.toLowerCase() === 'input') node.select();
}

module.exports = function createFocusGroup(options) {
  return new FocusGroup(options);
};

},{}]},{},[1]);
